Add a tool for the final release steps.

alexmv opened the pull request:
The final release steps are a bit fiddly, but are pretty well able to be automated.  Add a tool to do so.

I have not, obviously, made a release with this.  I've tested it to the extent I can.

I considered trying to do the follow-up steps (cherry-picking the changelog across, etc) but they started requiring more care, and automated code changes in shell, than I was comfortable with.

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

andersk commented on tools/build-release-tarball:
```diff
@@ -5,15 +5,27 @@ set -eu
-args="$(getopt -o '' -l 'help' -n "$0" -- "$@")"
-eval "set -- $args"
+args="$(getopt -o '' --long 'help,tag' -n "$0" -- "$@")"
+eval set -- "$args"
```
It was clearer not to rely on `eval` taking multiple arguments, because all does with them is smash them together with spaces. It doesn’t do any quoting. Interpolating `$args` into this command is only safe because `getopt` is responsible for quoting the arguments appropriately.

alexmv replied:
Thanks for catching that -- reverted.  That was a hunk leftover from when I was slowly diagnosing that BSD `getopt` is very much not GNU `getopt`, which was making running this on my local OSX machine very confusing.

timabbott commented:
I don't think I've used the option to have `build-release-tarball` create a tag (I always create it manually), but what key would be using for signing signed tags?

andersk commented:
It’s controlled by the `tag.gpgSign` Git configuration variable ([documentation](https://git-scm.com/docs/git-tag#Documentation/git-tag.txt--s)).

There’s also the question of which key we *want* to use, which is maybe what you meant.

alexmv commented:
I've dropped the pieces that generate a signed tag, as those need more discussion.  Based on [this czo conversation](https://chat.zulip.org/#narrow/stream/31-production-help/topic/upgrade.20to.205.2E3/near/1395143), I've added a heuristic that anything that looks like a formal release needs to have `--tag` set -- as well as a third commit which double-checks the version file that is produced.  I think the third commit is unnecessary given the other two (I can't imagine how it would fail), but certainly can't hurt.

andersk commented on tools/build-release-tarball:
```diff
@@ -27,13 +39,23 @@ if [ -z "${1:-}" ]; then
 fi
 
 version="$1"
+if [[ "$version" =~ ^[0-9]\.[0-9]+$ ]]; then
```
So we care about 4.11 but not 11.4? I’d just do `^[0-9.]*$` (this condition is in negative position and should err on the loose side, not the strict side).

Could join this with the following line using `&&`.

andersk commented on tools/build-release-tarball:
```diff
@@ -62,6 +84,18 @@ git checkout-index -f -a --prefix "$TMPDIR/$prefix/"
+        exit 1
+    fi
+    if [ "$base_version" != "$version" ]; then
+        echo "Version mismatch in generated version -- expected base version to be $base_version, got $base_version!"
```
`got $base_version` → `got $version`.

This will always fail with our current procedure, since the commit hasn’t yet been pushed upstream.

alexmv replied:
`expected base version to be $base_version` → `expected base version to be $version`, but you're right this will fail -- my testing I was doing wasn't with any local commits.

I think our options are:
1. Remove this check -- all of our previous releases have had a second line which is a previous commit, and have claimed in the version to be a fork from that version ~1 commit prior.  We change the logic for `is_fork` to say it isn't a fork if it's a `\d+\.\d+` tag.
2. Change `cache-zulip-git-version` to write the second line as the tag if the git describe was a tag, effectively moving the logic from (1) into the `zulip-git-version` file.
3. Change the release ordering to tag, push, build tarball, upload tarball -- which makes the `--tag` argument this PR adds impossible, unless `--tag` with a release-ish number also does a push.  Which feels wrong, in that that's not "building a release tarball" anymore, that's "doing a release."  So instead of this PR, we make a more general "release" tool, which does the equivalent of `git tag $version && git push && build-release-tarball $version && upload-release` to automate all of those together.

(3) seems like the best bet.

andersk replied:
Yeah, seems reasonable. A potential downside is that a `build-release-tarball` failure would burn a version number, but I trust our CI to make that unlikely.

andersk replied:
I guess we could get away with pushing just the commit up front, not the tag.

alexmv replied:
Repushed with a `tools/release` to do this.

timabbott commented on tools/release:
```diff
@@ -0,0 +1,124 @@
+    echo "Did not find expected $TARBALL!"
+    exit 1
+fi
+
```
Should we add a pause point here when you can hit enter to proceed with uploading or `ctrl+C` to cancel? The type of failure mode I'm imagining is you notice the changelog didn't format properly, or something.

timabbott commented on tools/release:
```diff
@@ -0,0 +1,124 @@
+# Upload to Github
+gh release create "$version" \
+    --title "Zulip Server $version" \
+    --notes-file <(echo "$changelog") \
```
We may need to do some reformatting on the changelog the first time or two, but I guess that's editable after publishing, so seems reasonable.
